### PR TITLE
 Add IL verification with ILVerify in addition to PEVerify

### PR DESCRIPTION
**Pull request by jcouv**

### CODE REVIEW

File: src/Compilers/Test/Core/Microsoft.CodeAnalysis.Test.Utilities.csproj
Comment (jaredpar) [inline src/Compilers/Test/Core/Microsoft.CodeAnalysis.Test.Utilities.csproj:109]: Why are we only referencing this for `net472`? Would expect ILVerify to work on all target frameworks.

File: src/Compilers/Test/Core/Platform/CoreClr/CoreCLRRuntimeEnvironment.cs
Comment (jaredpar) [inline src/Compilers/Test/Core/Platform/CoreClr/CoreCLRRuntimeEnvironment.cs:124]: Is it possible that you need simple names when dealing with non-strong named compilations and full names when dealing with strong named compilations?

File: src/Compilers/Test/Core/Platform/CoreClr/CoreCLRRuntimeEnvironment.cs
Comment (jaredpar) [inline src/Compilers/Test/Core/Platform/CoreClr/CoreCLRRuntimeEnvironment.cs:145]: Can probably delete that comment as a part of this change

File: src/Compilers/Test/Core/Platform/CoreClr/CoreCLRRuntimeEnvironment.cs
Comment (jaredpar) [inline src/Compilers/Test/Core/Platform/CoreClr/CoreCLRRuntimeEnvironment.cs:113]: The `CoreCLRRuntimeEnvironment` class comes into play when we are executing our tests on .NET Core vs. NET Framework. The ILVerify toolchain though should be runnable on any IL. Or at worst should be runnable when compiling for .NET Core. What the test compiles for and what runtime the test executes on are different concepts.

That made me a bit surprised to see this type here. My intuition was that ILVerification would be a general verification pass we could run independent of which runtime our tests were executing on. 

Am I missing something here?

File: src/Compilers/Test/Core/Platform/CoreClr/CoreCLRRuntimeEnvironment.cs
Comment (jaredpar) [inline src/Compilers/Test/Core/Platform/CoreClr/CoreCLRRuntimeEnvironment.cs:202]: Curious: our previous verifaction was mostly yes / no where this implementation is more looking for the exact failure case. Is there a reason you chose this route? I worry a bit that we're using messages to infer error codes. It would seem more natural to push strongly typed error codes into ILVerify if that was our desire here.

Comment (jcouv) [reply src/Compilers/Test/Core/Platform/CoreClr/CoreCLRRuntimeEnvironment.cs:202]: The main reason I thought it useful to break down the kinds of failures is that some of those are due to limitations in ILVerify or to the verification rules. Having a breakdown of the failures would help us inform where to improve ILVerify.
If you think we should stick with pass vs. skip, that's good feedback since it will much simplify the PR

File: src/Compilers/Test/Core/Platform/CoreClr/CoreCLRRuntimeEnvironment.cs
Comment (jcouv) [reply src/Compilers/Test/Core/Platform/CoreClr/CoreCLRRuntimeEnvironment.cs:113]: PEVerify is in `DesktopRuntimeEnvironment` so I'd followed that model, but I think you're right: IL verification likely can be done in `CompilationVerifier` instead of `CoreCLRRuntimeEnvironment`/`DesktopRuntimeEnvironment`. It just needs the emitted bits and the bits for all the dependencies. I'll take a look at that.

We can discuss whether ILVerify should also be run on desktop. I had it that way in an earlier commit, but then ran into some hurdles (should be possible to resolve, but didn't want to block progress on main goal).

File: src/Compilers/Test/Core/Microsoft.CodeAnalysis.Test.Utilities.csproj
Comment (jcouv) [reply src/Compilers/Test/Core/Microsoft.CodeAnalysis.Test.Utilities.csproj:109]: I thought I'd reverted that. I was trying to solve the problem with strong name, assuming that the problem was only occurring for net472.

File: src/Compilers/Test/Core/Platform/CoreClr/CoreCLRRuntimeEnvironment.cs
Comment (jaredpar) [reply src/Compilers/Test/Core/Platform/CoreClr/CoreCLRRuntimeEnvironment.cs:202]: For the initial PR I would favor sticking with our existing pass / fail semantics. If we have a desire to make it more complicated I think that's reasonable but I would start by pushing ILVerify to produce diagnostics that are more strongly typed (an `enum`, error code, etc ...).

File: src/Compilers/Test/Core/Microsoft.CodeAnalysis.Test.Utilities.csproj
Comment (jaredpar) [reply src/Compilers/Test/Core/Microsoft.CodeAnalysis.Test.Utilities.csproj:109]: Have we followed up with the team yet about why it's not strong named? That is our existing recommendation so it's odd that it is not being followed. I'm happy to chat with the owners about this.

File: src/Compilers/Test/Core/CompilationVerifier.cs
Comment (jcouv) [inline src/Compilers/Test/Core/CompilationVerifier.cs:271]: Note to self: Add a note/issue to get the module bits without loading them in memory

File: src/Compilers/Test/Core/ExceptionHelper.cs
Comment (jaredpar) [inline src/Compilers/Test/Core/ExceptionHelper.cs:32]: Nit: generally they use the first two letters are upper case

```suggestion
            string tool = isIlVerify ? "ILVerify" : "PEVerify";
```

File: src/Compilers/Test/Core/Microsoft.CodeAnalysis.Test.Utilities.csproj
Comment (jaredpar) [inline src/Compilers/Test/Core/Microsoft.CodeAnalysis.Test.Utilities.csproj:108]:
```suggestion
```

File: src/Compilers/Test/Core/Platform/CoreClr/CoreCLRRuntimeEnvironment.cs
Comment (jaredpar) [inline src/Compilers/Test/Core/Platform/CoreClr/CoreCLRRuntimeEnvironment.cs:62]: +1 this is the best way to approach this problem

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenExprLambdaTests.cs
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenExprLambdaTests.cs:3120]: One part I'm having trouble deciding on is the comments for the failures. There are effectively three reasons that we have failures:

1. Legitimate failures: consider there are several cases where we run verification with less than the full assembly set. For example missing corelib. Verification fails there essentially by design. 
2. Bugs or limitations in PEVerify: due to the servicing nature of PEVerify these will never be fixed
3. Bugs or limitations in ILVerify: these can be prioritize and fixed

Can't decide if having comments tracking all three cases is worth it. For some reason it just stands out when I'm reviewing the change. Possible that is because this is calling out the ~100 or so times this happens out of the hundreds of thousands of tests that we have hence I'm just looking at the worst case. Means I won't notice in the typical flow of PRs through the repo. 

Don't feel strongly about this but wanted to hear your thoughts a bit.

File: src/Compilers/Test/Core/CompilationVerifier.cs
Comment (jcouv) [reply src/Compilers/Test/Core/CompilationVerifier.cs:241]: Right. The issue is that `ILVerify.ResolverBase.ResolveCore(string simpleName)` gives a simple name.
So tests that deal with duplicate simple names will need to skip ILVerify for the moment.

File: src/Compilers/Test/Core/CompilationVerifier.cs
Comment (jcouv) [reply src/Compilers/Test/Core/CompilationVerifier.cs:251]: Good catch. In an earlier iteration, I was storing "name -> default" into the map for duplicates.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenExprLambdaTests.cs
Comment (jcouv) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenExprLambdaTests.cs:3120]: Feels similar to me to leaving a comment when we skip a test. It helps clarify whether we're in situation 1 (test prohibits validation, and clarify that it's not a case where the compiler emitted something bad) or in situations 2 or 3.

Comments for situation 3 are useful to prioritize improving ILVerify (we can do a search for `// ILVerify:` in tests).
I don't much care for comments for situation 2 either (since PEVerify won't get updated), other than clarifying that it's not situation 1.

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for a second review. From discussion with Jared it would be good to have a second review even though this is a test-only change.
Note: the CI failures are expected as new tests have been added to `main`. I'll merge them when the PR is otherwise ready to merge.

**Review by jaredpar — APPROVED** [review]

File: src/Compilers/Test/Core/CompilationVerifier.cs
Comment (cston) [inline src/Compilers/Test/Core/CompilationVerifier.cs:239]:
```suggestion
                    if (_imagesByName.ContainsKey(name))
```

File: src/Compilers/Test/Core/CompilationVerifier.cs
Comment (cston) [inline src/Compilers/Test/Core/CompilationVerifier.cs:25]: Consider sorting usings.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs
Comment (cston) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs:104]: >Assert.Equal("C", model.GetTypeInfo(right).ConvertedType.ToTestDisplayString()); [](http://example.com/codeflow?start=16&length=80)

Is this change necessary?

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs
Comment (cston) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs:13]: >System.Reflection.PortableExecutable [](http://example.com/codeflow?start=6&length=36)

Is this used?

File: src/Compilers/CSharp/Test/Emit2/Attributes/AttributeTests_RefReadOnly.cs
Comment (cston) [inline src/Compilers/CSharp/Test/Emit2/Attributes/AttributeTests_RefReadOnly.cs:1163]: Formatting appears to be incorrect. The change can be reverted.

File: src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs:1301]: >ILVerify: Unrecognized arguments for delegate .ctor. [](http://example.com/codeflow?start=19&length=52)

What does this error refer to?

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs:2394]: >Verification.FailsIlVerify [](http://example.com/codeflow?start=69&length=26)

What are the failures for tests in the file?

File: src/Compilers/CSharp/Test/Symbol/Symbols/ExtensionMethodTests.cs
Comment (cston) [inline src/Compilers/CSharp/Test/Symbol/Symbols/ExtensionMethodTests.cs:297]: >Verification.FailsIlVerify [](http://example.com/codeflow?start=14&length=26)

What are the failures for tests in this file?

File: src/Compilers/Test/Core/CompilationVerifier.cs
Comment (cston) [inline src/Compilers/Test/Core/CompilationVerifier.cs:282]: >throw new Exception [](http://example.com/codeflow?start=20&length=19)

Is it necessary to throw exceptions here if we're catching the exceptions in the same method? (Or are we also expecting exceptions from `verify.Verify()`?

Comment (jcouv) [reply src/Compilers/Test/Core/CompilationVerifier.cs:282]: Yes, `verifier.Verify()` throws exceptions too

File: src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs:1301]: Unfortunately, that's all the details from ILVerify at this point.

File: src/Compilers/Test/Core/Platform/CoreClr/CoreCLRRuntimeEnvironment.cs
Comment (cston) [inline src/Compilers/Test/Core/Platform/CoreClr/CoreCLRRuntimeEnvironment.cs:22]: Consider sorting.

File: src/Compilers/Test/Core/Platform/CoreClr/CoreCLRRuntimeEnvironment.cs
Comment (cston) [inline src/Compilers/Test/Core/Platform/CoreClr/CoreCLRRuntimeEnvironment.cs:116]: >Verify [](http://example.com/codeflow?start=20&length=6)

Is this method needed?

Comment (jcouv) [reply src/Compilers/Test/Core/Platform/CoreClr/CoreCLRRuntimeEnvironment.cs:116]: Yes, it implements `IRuntimeEnvironment`

File: src/Compilers/Test/Core/CommonTestBase.cs
Comment (cston) [inline src/Compilers/Test/Core/CommonTestBase.cs:33]: Consider using "PE" and "IL" capitalization.

File: src/Compilers/CSharp/Test/Symbol/Symbols/ExtensionMethodTests.cs
Comment (cston) [inline src/Compilers/CSharp/Test/Symbol/Symbols/ExtensionMethodTests.cs:293]: >ILVerify: Unrecognized arguments for delegate .ctor. { Offset = 21 } [](http://example.com/codeflow?start=15&length=68)

Is this error reported for a delegate created from an extension method? And does it indicate a bug in Roslyn or in ILVerify?

Comment (jcouv) [reply src/Compilers/CSharp/Test/Symbol/Symbols/ExtensionMethodTests.cs:293]: Since PEVerify passes, I think it's an ILVerify limitation.
The IL looks fine as far as I can tell:
```
.class private auto ansi abstract sealed beforefieldinit Program
	extends [netstandard]System.Object
{
	.custom instance void [netstandard]System.Runtime.CompilerServices.ExtensionAttribute::.ctor() = (
		01 00 00 00
	)
	// Methods
	.method private hidebysig static 
		void Main () cil managed 
	{
		// Method begins at RVA 0x2050
		// Code size 70 (0x46)
		.maxstack 2
		.entrypoint
		IL_0000: ldstr "ABC"
		IL_0005: call void Program::Goo(string)
		IL_000a: ldstr "123"
		IL_000f: ldftn void Program::Goo(string)
		IL_0015: newobj instance void [netstandard]System.Action::.ctor(object, native int)
		IL_001a: dup
		IL_001b: callvirt instance void [netstandard]System.Action::Invoke()
		IL_0020: ldftn instance void [netstandard]System.Action::Invoke()
		IL_0026: newobj instance void [netstandard]System.Action::.ctor(object, native int)
		IL_002b: callvirt instance void [netstandard]System.Action::Invoke()
		IL_0030: ldstr "xyz"
		IL_0035: ldftn void Program::Goo(string)
		IL_003b: newobj instance void [netstandard]System.Action::.ctor(object, native int)
		IL_0040: callvirt instance void [netstandard]System.Action::Invoke()
		IL_0045: ret
	} // end of method Program::Main
	.method private hidebysig static 
		void Goo (
			string x
		) cil managed 
	{
		.custom instance void [netstandard]System.Runtime.CompilerServices.ExtensionAttribute::.ctor() = (
			01 00 00 00
		)
		// Method begins at RVA 0x20a2
		// Code size 7 (0x7)
		.maxstack 8
		IL_0000: ldarg.0
		IL_0001: call void [netstandard]System.Console::WriteLine(string)
		IL_0006: ret
	} // end of method Program::Goo
} // end of class Program
```

Comment (jcouv) [reply src/Compilers/CSharp/Test/Symbol/Symbols/ExtensionMethodTests.cs:293]: Confirmed the issue.
ILVerify check whether `Program.Goo(string)` can be assigned to delegate type `Action` and fails. It doesn't account for `this`/extension scenario.

```
        void CheckIsDelegateAssignable(MethodDesc ftn, TypeDesc delegateType)
        {
            if (!IsDelegateAssignable(ftn, delegateType))
                VerificationError(VerifierError.DelegateCtor);
        }
        
        bool IsDelegateAssignable(MethodDesc ftn, TypeDesc delegateType)
        {
            var invokeMethod = delegateType.GetMethod("Invoke", null);
            if (invokeMethod == null)
                return false;

            var ftnSignature = ftn.Signature;
            var delegateSignature = invokeMethod.Signature;

            // Compare calling convention ignoring distinction between static and instance
            if ((ftnSignature.Flags & ~MethodSignatureFlags.Static) != (delegateSignature.Flags & ~MethodSignatureFlags.Static))
                return false;

            // Compare signature parameters
            if (ftnSignature.Length != delegateSignature.Length)
                return false;

            for (int i = 0; i < ftnSignature.Length; i++)
            {
                if (!IsAssignable(delegateSignature[i], ftnSignature[i]))
                    return false;
            }

            // Compare return type
            return IsAssignable(ftnSignature.ReturnType, delegateSignature.ReturnType);
        }
```

Comment (jcouv) [reply src/Compilers/CSharp/Test/Symbol/Symbols/ExtensionMethodTests.cs:293]: Filed https://github.com/dotnet/runtime/issues/65010

File: src/Compilers/Test/Core/CompilationVerifier.cs
Comment (cston) [inline src/Compilers/Test/Core/CompilationVerifier.cs:282]: It looks like the `catch` block is unnecessary.

Comment (jaredpar) [reply src/Compilers/Test/Core/CompilationVerifier.cs:282]: Great `catch` 😄

**Review by cston — APPROVED** [review]

File: src/Compilers/Test/Core/CompilationVerifier.cs
Comment (cston) [inline src/Compilers/Test/Core/CompilationVerifier.cs:299]: >Exception [](http://example.com/codeflow?start=19&length=9)

By catching all exceptions, aren't we potentially hiding bugs? For instance, `Microsoft.CodeAnalysis.VisualBasic.UnitTests.CodeGenTests.TestGetObjectValueCalls()` seems to fail `ILVerify()` because of a `NullReferenceException`.

File: src/Compilers/Test/Core/CompilationVerifier.cs
Comment (cston) [inline src/Compilers/Test/Core/CompilationVerifier.cs:288]: >throw new Exception [](http://example.com/codeflow?start=20&length=19)

Won't the outer `catch` clause simply `return` without failing with `if ((verification & Verification.FailsILVerify) != 0) { return ; }`?

File: src/Compilers/Test/Core/CompilationVerifier.cs
Comment (cston) [inline src/Compilers/Test/Core/CompilationVerifier.cs:276]: >var mscorlibModule = _allModuleData.Single(m => m.IsCorLib); [](http://example.com/codeflow?start=16&length=60)

Consider moving outside of `try { } catch { }`. (It looks like this throws an `InvalidOperationException` from `Microsoft.CodeAnalysis.CSharp.UnitTests.UseSiteErrorTests.MissingTypeKindBasisTypes()`.

File: src/Compilers/Test/Core/CompilationVerifier.cs
Comment (jcouv) [reply src/Compilers/Test/Core/CompilationVerifier.cs]: There was indeed a bug here. Thanks. Fixed by reducing usage of exceptions for failure scenarios.

File: src/Compilers/Test/Core/CompilationVerifier.cs
Comment (jcouv) [reply src/Compilers/Test/Core/CompilationVerifier.cs]: From previous discussion with Jared, we don't distinguish between the different failure modes of ILVerify. Either it succeeds or it fails. That's why I added a comment to the test:
```
            ' ILVerify null ref
            ' Tracked by https//github.com/dotnet/roslyn/issues/58652
```

File: src/Compilers/Test/Core/CompilationVerifier.cs
Comment (cston) [inline src/Compilers/Test/Core/CompilationVerifier.cs:238]: >= new Dictionary<string, ImmutableArray<byte>>(StringComparer.OrdinalIgnoreCase) [](http://example.com/codeflow?start=84&length=80)

The initial value is overwritten in the constructor.

File: src/Compilers/Test/Core/CompilationVerifier.cs
Comment (cston) [inline src/Compilers/Test/Core/CompilationVerifier.cs:291]: >mscorlibModule.SimpleName [](http://example.com/codeflow?start=61&length=25)

If `(verification & Verification.FailsILVerify) == 0`, `mscorlibModule` may be null.

File: src/Compilers/Test/Core/CompilationVerifier.cs
Comment (cston) [inline src/Compilers/Test/Core/CompilationVerifier.cs:316]: Consider moving this outside the `try { } catch { }`.

**Review by cston — APPROVED** [review]

File: src/Compilers/Test/Core/CompilationVerifier.cs
Comment (jkotas) [reply src/Compilers/Test/Core/CompilationVerifier.cs]: FYI: https://github.com/dotnet/runtime/issues/65573 is proposing passing the full assembly name to the resolver.

**Pull request merged**
